drivers/spi/spi_iproc_qspi.c

```diff
@@ -0,0 +1,433 @@
+			spi_context_update_tx(&data->ctx, 1, data->ctx.tx_len);
+			memcpy(tx + tx_len, ctx->tx_buf, ctx->tx_len);
+
+			tx_len += ctx->tx_len;
+		}
+
+		ret = mspi_transfer(dev, tx, tx_len, rx, rx_len);
+		spi_context_update_tx(ctx, 1, ctx->tx_len);
+	} else {
+		/* update tx rx ctx buf ptrs and get nxt data buf pointers */
+		spi_context_update_tx(ctx, 1, ctx->tx_len);
+		spi_context_update_rx(&data->ctx, 1, data->ctx.rx_len);
+		rx = ctx->rx_buf;
+		rx_len = ctx->rx_len;
+
+		ret = mspi_transfer(dev, tx, tx_len, rx, rx_len);
+		/* update tx also to indicate end of xfer */
+		spi_context_update_tx(ctx, 1, ctx->tx_len);
+		spi_context_update_rx(ctx, 1, rx_len);
+	}
+
+	return ret;
+}
+
+static bool iproc_qspi_transfer_ongoing(struct iproc_qspi_data *data)
+{
+	return spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx);
+}
+
+static int iproc_qspi_setup(const struct device *dev, const struct spi_config *spi_cfg)
+{
+	uint32_t val = 0x0;
+
+	/* MSPI: Basic hardware initialization */
+	sys_write32(0, DEV_MSPI_BASE(dev) + MSPI_SPCR1_LSB);
+	sys_write32(0, DEV_MSPI_BASE(dev) + MSPI_SPCR1_MSB);
+	sys_write32(0, DEV_MSPI_BASE(dev) + MSPI_NEWQP);
+	sys_write32(0, DEV_MSPI_BASE(dev) + MSPI_ENDQP);
+	sys_write32(0, DEV_MSPI_BASE(dev) + MSPI_SPCR2);
+
+	/* MSPI: SCK configuration */
+	sys_write32(SPBR_MIN, DEV_MSPI_BASE(dev) + MSPI_SPCR0_LSB);
+
+	if (SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_CPOL) {
+		val |= BIT(MSPI_SPCR0_MSB_CPOL_SHIFT);
+	}
+
+	if (SPI_MODE_GET(spi_cfg->operation) & SPI_MODE_CPHA) {
+		val |= BIT(MSPI_SPCR0_MSB_CPHA_SHIFT);
+	}
+
+	val |= (SPI_WORD_SIZE_GET(spi_cfg->operation) << MSPI_SPCR0_MSB_BITS_SHIFT);
+
+	val |= BIT(MSPI_SPCR0_MSB_MSTR_SHIFT);
+
+	/* MSPI: Mode configuration */
+	sys_write32(val, DEV_MSPI_BASE(dev) + MSPI_SPCR0_MSB);
+
+	return 0;
+}
+
+static int iproc_qspi_configure(const struct device *dev, const struct spi_config *spi_cfg)
+{
+	struct iproc_qspi_data *data = DEV_DATA(dev);
+
+	if (spi_context_configured(&data->ctx, spi_cfg)) {
+		/* This configuration is already in use */
+		return 0;
+	}
+
+	if (SPI_WORD_SIZE_GET(spi_cfg->operation) != 8) {
+		return -ENOTSUP;
+	}
+
+	iproc_qspi_setup(dev, spi_cfg);
+
+	/* At this point, it's mandatory to set this on the context! */
+	data->ctx.config = spi_cfg;
+
+	LOG_DBG("Installed config %p, slave %u", spi_cfg, spi_cfg->slave);
+
+	return 0;
+}
+
+static int transceive(const struct device *dev, const struct spi_config *spi_cfg,
+		      const struct spi_buf_set *tx_bufs, const struct spi_buf_set *rx_bufs,
+		      bool asynchronous, spi_callback_t cb, void *userdata)
+{
+	int ret;
+	struct iproc_qspi_data *data = DEV_DATA(dev);
+
+	spi_context_lock(&data->ctx, asynchronous, cb, userdata, spi_cfg);
+
+	ret = iproc_qspi_configure(dev, spi_cfg);
+	if (ret) {
+		goto release;
+	}
+
+	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
+
+	spi_context_cs_control(&data->ctx, true);
+
+	do {
+		ret = iproc_qspi_xfer(dev);
```

Comment:
I don't think such advanced controller does not support interrupt mode. 

Please, switch to interrupt mode. This is a mandatory requirement.

Reply:
Unfortunately this is what we (Meta) have, I think it provides a good starting point to whoever is going to use it, PRs are welcomed to add more features - its open sourced.